src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "pandaproxy/kafka_client_cache.h"
+
+#include "pandaproxy/client_cache_error.h"
+#include "pandaproxy/logger.h"
+
+#include <chrono>
+
+using namespace std::chrono_literals;
+
+namespace pandaproxy {
+
+static constexpr auto clean_timer_period = 10s;
+
+kafka_client_cache::kafka_client_cache(
+  YAML::Node const& cfg,
+  size_t max_size,
+  std::vector<config::broker_authn_endpoint> kafka_api,
+  model::timestamp::type keep_alive)
+  : _config{cfg}
+  , _cache_max_size{max_size}
+  , _kafka_has_sasl{false}
+  , _keep_alive{keep_alive} {
+    // Is there a Kafka listener with SASL enabled?
+    auto ep_it = std::find_if(
+      kafka_api.begin(),
+      kafka_api.end(),
+      [](const config::broker_authn_endpoint& ep) {
+          return ep.authn_method == config::broker_authn_method::sasl;
+      });
+
+    _kafka_has_sasl = ep_it != kafka_api.end();
+
+    _clean_timer.set_callback([this] {
+        // Do not need a gate since there are limited items
+        // in the cache. Therefore, relative to other design,
+        // garbage collection is inexpensive here.
+        clean_stale_clients();
```

Comment:
There may be a synchronization issue here. Consider the situation where:
1. Client1 is fetched from the cache but the handler has not yet used the client to service a request
2. Client clean-up is invoked on client1
3. The handler attempts to service the request but client1 was removed already

The solution would be to use a gate/mutex on each kafka client cache.